flyteplugins/go/tasks/plugins/k8s/spark/spark.go

```diff
@@ -165,57 +109,158 @@ func (sparkResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsCo
+			sparkConfig[sparkOpConfig.SparkDriverCoreLimitKey] = sparkConfig["spark.driver.cores"]
 		}
 	}
 
-	if len(sparkConfig["spark.kubernetes.executor.limit.cores"]) == 0 {
+	if len(sparkConfig[sparkOpConfig.SparkExecutorCoreLimitKey]) == 0 {
 		// spark.kubernetes.executor.request.cores takes precedence over spark.executor.cores
-		if len(sparkConfig["spark.kubernetes.executor.request.cores"]) != 0 {
-			sparkConfig["spark.kubernetes.executor.limit.cores"] = sparkConfig["spark.kubernetes.executor.request.cores"]
+		if len(sparkConfig[sparkOpConfig.SparkExecutorCoreRequestKey]) != 0 {
+			sparkConfig[sparkOpConfig.SparkExecutorCoreLimitKey] = sparkConfig[sparkOpConfig.SparkExecutorCoreRequestKey]
 		} else if len(sparkConfig["spark.executor.cores"]) != 0 {
-			sparkConfig["spark.kubernetes.executor.limit.cores"] = sparkConfig["spark.executor.cores"]
+			sparkConfig[sparkOpConfig.SparkExecutorCoreLimitKey] = sparkConfig["spark.executor.cores"]
 		}
 	}
 
 	sparkConfig["spark.kubernetes.executor.podNamePrefix"] = taskCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName()
 	sparkConfig["spark.kubernetes.driverEnv.FLYTE_START_TIME"] = strconv.FormatInt(time.Now().UnixNano()/1000000, 10)
 
-	// Add driver/executor defaults to CRD Driver/Executor Spec as well.
-	cores, err := strconv.ParseInt(sparkConfig["spark.driver.cores"], 10, 32)
-	if err == nil {
-		driverSpec.Cores = intPtr(int32(cores))
+	return sparkConfig
+}
+
+func serviceAccountName(metadata pluginsCore.TaskExecutionMetadata) string {
+	name := flytek8s.GetServiceAccountNameFromTaskExecutionMetadata(metadata)
+	if len(name) == 0 {
+		name = sparkTaskType
 	}
-	driverSpec.Memory = strPtr(sparkConfig["spark.driver.memory"])
+	return name
+}
 
-	execCores, err := strconv.ParseInt(sparkConfig["spark.executor.cores"], 10, 32)
-	if err == nil {
-		executorSpec.Cores = intPtr(int32(execCores))
+func createSparkPodSpec(ctx context.Context, taskCtx pluginsCore.TaskExecutionContext, podSpec *v1.PodSpec, container *v1.Container) (
+	*sparkOp.SparkPodSpec, error) {
+	annotations := utils.UnionMaps(config.GetK8sPluginConfig().DefaultAnnotations, utils.CopyMap(taskCtx.TaskExecutionMetadata().GetAnnotations()))
+	labels := utils.UnionMaps(config.GetK8sPluginConfig().DefaultLabels, utils.CopyMap(taskCtx.TaskExecutionMetadata().GetLabels()))
+
+	sparkEnvVars := make(map[string]string)
+	for _, envVar := range container.Env {
+		sparkEnvVars[envVar.Name] = envVar.Value
 	}
+	sparkEnvVars["FLYTE_MAX_ATTEMPTS"] = strconv.Itoa(int(taskCtx.TaskExecutionMetadata().GetMaxAttempts()))
+
+	spec := sparkOp.SparkPodSpec{
+		Affinity:         podSpec.Affinity,
+		Annotations:      annotations,
+		Labels:           labels,
+		EnvVars:          sparkEnvVars,
+		Image:            &container.Image,
+		SecurityContenxt: podSpec.SecurityContext.DeepCopy(),
+		DNSConfig:        podSpec.DNSConfig.DeepCopy(),
+		Tolerations:      podSpec.Tolerations,
+		SchedulerName:    &config.GetK8sPluginConfig().SchedulerName,
```

Comment:
Can we use `podSpec.SchedulerName` here? Similarly for `HostNetwork`. These should be correctly set in `ToK8sPodSpec`.